### PR TITLE
Add DependencyInjection event source

### PR DESCRIPTION
**Pull request by pakrym**
With one event that dumps call site trees that are being created

### CODE REVIEW

File: src/DependencyInjection/DI/src/CallSiteFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        protected override object VisitServiceScopeFactory(ServiceScopeFactoryCallSite serviceScopeFactoryCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("<ScopeFactory>");
+
+            return null;
+        }
+
+        protected override object VisitIEnumerable(IEnumerableCallSite enumerableCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("IEnumerable<{0}> (size {1})", enumerableCallSite.ItemType, enumerableCallSite.ServiceCallSites.Length);
```
Comment (davidfowl) [inline src/DependencyInjection/DI/src/CallSiteFormatter.cs]: nit: Remove size, just put the number

File: src/DependencyInjection/DI/src/CallSiteFormatter.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            {
+                VisitCallSite(parameter, newContext);
+            }
+
+            return null;
+        }
+
+        protected override object VisitConstant(ConstantCallSite constantCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("{0}", constantCallSite.DefaultValue ?? "<null>");
```
Comment (anurse) [inline src/DependencyInjection/DI/src/CallSiteFormatter.cs]: Do we expect this value to be formattable? Most of the time these are user-provided instances of services, right? 

Perhaps just `<Constant: [TypeName]>` would be more appropriate.

File: src/DependencyInjection/DI/src/CallSiteFormatter.cs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            {
+                VisitCallSite(item, newContext);
+            }
+
+            return null;
+        }
+
+        protected override object VisitFactory(FactoryCallSite factoryCallSite, CallSiteFormatterContext argument)
+        {
+            argument.ContinueLine("Factory ({0})", factoryCallSite.Factory.Method);
```
Comment (anurse) [inline src/DependencyInjection/DI/src/CallSiteFormatter.cs]: The other "special token" values use `<>`, I think that might be appropriate here: `<Factory: [MethodName]>`

File: src/DependencyInjection/DI/src/CallSiteFormatter.cs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+            return null;
+        }
+
+        protected override object VisitDisposeCache(ServiceCallSite callSite, CallSiteFormatterContext argument)
+        {
+            argument.StartLine("DisposeCache ");
```
Comment (anurse) [inline src/DependencyInjection/DI/src/CallSiteFormatter.cs]: Are these decorators applied to call-sites?

Comment (anurse) [reply src/DependencyInjection/DI/src/CallSiteFormatter.cs]: Might be clearer if these were something like `[DisposeCache] `

File: src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+using System.Diagnostics.Tracing;
+using System.Runtime.CompilerServices;
+using Microsoft.Extensions.DependencyInjection.ServiceLookup;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    [EventSource(Name = "Microsoft-Extensions-DependencyInjection")]
+    internal sealed class DependencyInjectionEventSource : EventSource
+    {
+        public static readonly DependencyInjectionEventSource Instance = new DependencyInjectionEventSource();
```
Comment (anurse) [inline src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs]: nit: The idiomatic name of this is `Log`... not sure why, but it just is :)

https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventsource?view=netframework-4.7.2#examples

Not sure what other ASP.NET Core sources use though.

File: src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        [Event(1, Level = EventLevel.Verbose)]
+        private void CallSiteBuilt(string callSite)
+        {
+            WriteEvent(1,  callSite);
+        }
+
+        [NonEvent]
+        public void CallSiteBuilt(ServiceCallSite callSite)
+        {
+            if (IsEnabled())
```
Comment (anurse) [inline src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs]: Pass in the level here, to protect us when/if we add non-Verbose events.

File: src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+        // NOTE
+        // - The 'Start' and 'Stop' suffixes on the following event names have special meaning in EventSource. They
+        //   enable creating 'activities'.
+        //   For more information, take a look at the following blog post:
+        //   https://blogs.msdn.microsoft.com/vancem/2015/09/14/exploring-eventsource-activity-correlation-and-causation-features/
+        // - A stop event's event id must be next one after its start event.
+        // - Avoid renaming methods or parameters marked with EventAttribute. EventSource uses these to form the event object.
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (anurse) [inline src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs]: Any reason for no-inlining here?

File: src/DependencyInjection/DI/test/DependencyInjectionEventSourceTests.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+        [Fact]
+        public void EmitsCallSiteBuiltEvent()
+        {
+            // Arrange
+            var serviceCollection = new ServiceCollection();
+            serviceCollection.AddSingleton<FakeDisposeCallback>();
+            serviceCollection.AddTransient<IFakeOuterService, FakeDisposableCallbackOuterService>();
+            serviceCollection.AddSingleton<IFakeMultipleService, FakeDisposableCallbackInnerService>();
+            serviceCollection.AddScoped<IFakeMultipleService, FakeDisposableCallbackInnerService>();
```
Comment (anurse) [inline src/DependencyInjection/DI/test/DependencyInjectionEventSourceTests.cs]: Could you put a couple other types of call-sites in? A Factory and a Constant in particular would be nice.

File: src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        [Event(1, Level = EventLevel.Verbose)]
+        private void CallSiteBuilt(string callSite)
```
Comment (anurse) [inline src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs]: Is there anything we should promote to top-level metadata here? The service type name for example?

It's easier to query/filter events by payload values, so things that we think people will want to query should be direct parameters.

File: src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs (hunk 7 above)
Comment (pakrym) [reply src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs]: Copied from Kestrel.

File: src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs (hunk 6 above)
Comment (pakrym) [reply src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs]: `IsEnabled` doesn't take level.

File: src/DependencyInjection/DI/src/CallSiteFormatter.cs (hunk 2 above)
Comment (pakrym) [reply src/DependencyInjection/DI/src/CallSiteFormatter.cs]: `object.ToString()` defaults to type name anyway.

File: src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs (hunk 9 above)
Comment (pakrym) [reply src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs]: Added serviceType.

File: src/DependencyInjection/DI/test/DependencyInjectionEventSourceTests.cs (hunk 8 above)
Comment (pakrym) [reply src/DependencyInjection/DI/test/DependencyInjectionEventSourceTests.cs]: Done

**Comment by pakrym** [comment]
I changed the format to JSON. It's not as human readable but better for bulk analysis.

I'll switch to a real JSON writer when we get appropriate BCL version.

File: src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs (hunk 7 above)
Comment (pakrym) [reply src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs]: Removed.

File: src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs (hunk 5 above)
Comment (pakrym) [reply src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs]: Renamed.

File: src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs (hunk 6 above)
Comment (anurse) [reply src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs]: https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.tracing.eventsource.isenabled?view=netframework-4.7.2#System_Diagnostics_Tracing_EventSource_IsEnabled_System_Diagnostics_Tracing_EventLevel_System_Diagnostics_Tracing_EventKeywords_ ?

**Comment by anurse** [comment]
> I'll switch to a real JSON writer when we get appropriate BCL version.

File a bug to track this? Otherwise we'll just forget it :)

File: src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs (hunk 6 above)
Comment (pakrym) [reply src/DependencyInjection/DI/src/DependencyInjectionEventSource.cs]: Whoops.

**Pull request merged**